packages/core/src/routes/interaction/middleware/koa-interaction-hooks.ts

```diff
@@ -1,31 +1,61 @@
-import { trySafe } from '@silverhand/essentials';
+import { conditional, trySafe } from '@silverhand/essentials';
 import type { MiddlewareType } from 'koa';
 import type { IRouterParamContext } from 'koa-router';
 
-import type TenantContext from '#src/tenants/TenantContext.js';
+import type Libraries from '#src/tenants/Libraries.js';
 
 import { getInteractionStorage } from '../utils/interaction.js';
 
 import type { WithInteractionDetailsContext } from './koa-interaction-details.js';
 
+export type InteractionHookDetails = {
+  sessionId?: string;
+  userId?: string;
+  applicationId?: string;
+  userAgent?: string;
+};
+
+type AssignInteractionHookDetails = (details: Partial<InteractionHookDetails>) => void;
+
+export type WithInteractionHooksContext<
+  ContextT extends IRouterParamContext = IRouterParamContext
+> = ContextT & { assignInteractionHookDetails: AssignInteractionHookDetails };
+
 export default function koaInteractionHooks<
   StateT,
-  ContextT extends WithInteractionDetailsContext<IRouterParamContext>,
+  ContextT extends WithInteractionDetailsContext,
   ResponseT
 >({
-  provider,
-  libraries: {
-    hooks: { triggerInteractionHooksIfNeeded },
-  },
-}: TenantContext): MiddlewareType<StateT, ContextT, ResponseT> {
+  hooks: { triggerInteractionHooksIfNeeded },
+}: Libraries): MiddlewareType<StateT, WithInteractionHooksContext<ContextT>, ResponseT> {
   return async (ctx, next) => {
     const { event } = getInteractionStorage(ctx.interactionDetails.result);
 
+    const interactionHookDetails: InteractionHookDetails = {};
+
+    const {
+      interactionDetails: {
+        jti: sessionId,
+        params: { client_id },
+      },
+      header: { 'user-agent': userAgent },
+    } = ctx;
```

Comment:
i'm not sure this is a proper combination since:
- it create another data source.
- it is confusing that all those values are overridable via `interactionHookDetails` even if we the possibility of overriding is rare.
- it also change the time of fetching data like `jti` and `client_id`, from up-to-date data to cached data.

Reply:
especially `user-agent`, it's nothing related to the interaction

Reply:
how about you just remove all these initial values (including `event`) and assign them all inside `submitInteraction()`.